### PR TITLE
ARROW-12069: [C++][Gandiva] Implement IN expressions for Decimal type

### PR DESCRIPTION
**Pull request by jvictorhuguenin**
Implement support to IN expressions at Arrow's decimals field independently of its precision or scale.

### CODE REVIEW

**Comment by frank400** [comment]
@projjal, I fixed the builds, can you please review it now?

File: cpp/src/gandiva/decimal_scalar.h [hunk 1]
```diff
@@ -55,3 +55,19 @@ class DecimalScalar128 : public BasicDecimalScalar128 {
...
+struct hash<gandiva::DecimalScalar128> {
+  std::size_t operator()(gandiva::DecimalScalar128 const& s) const noexcept {
+    arrow::BasicDecimal128 dvalue(s.value());
+    std::size_t h0 = std::hash<int64_t>{}(dvalue.high_bits());
+    std::size_t h1 = std::hash<uint64_t>{}(dvalue.low_bits());
+
+    std::size_t h2 = std::hash<int32_t>{}(s.precision());
+    std::size_t h3 = std::hash<int32_t>{}(s.scale());
+
+    return (((h0 ^ (h1 << 1) >> 1) ^ (h2 << 1) >> 1) ^ h3 << 1);
```
Comment (projjal) [inline cpp/src/gandiva/decimal_scalar.h:70]: why not use arrow:util:hash_combine

Comment (frank400) [reply cpp/src/gandiva/decimal_scalar.h]: I didn't know about it yet, can you just check if I should hash the values before I pass then to the hash_combine function?

Comment (projjal) [reply cpp/src/gandiva/decimal_scalar.h]: No. Take a look at https://github.com/apache/arrow/blob/master/cpp/src/gandiva/filter.cc#L48

File: cpp/src/gandiva/tests/in_expr_test.cc [hunk 2]
```diff
@@ -34,6 +35,19 @@ class TestIn : public ::testing::Test {
  protected:
   arrow::MemoryPool* pool_;
 };
+std::vector<Decimal128> MakeDecimalVector(std::vector<std::string> values) {
+  std::vector<arrow::Decimal128> ret;
+  for (auto str : values) {
+    Decimal128 str_value;
+    int32_t str_precision;
+    int32_t str_scale;
```
Comment (projjal) [inline cpp/src/gandiva/tests/in_expr_test.cc:43]: decimal_value, decimal_precision, decimal_scale?

File: cpp/src/gandiva/decimal_scalar.h (hunk 1 above)
Comment (frank400) [reply cpp/src/gandiva/decimal_scalar.h]: Ok, I already had saw it, gonna change it right now

File: cpp/src/gandiva/tests/in_expr_test.cc (hunk 2 above)
Comment (frank400) [reply cpp/src/gandiva/tests/in_expr_test.cc]: changed

**Review by projjal — APPROVED** [review]

**Review by projjal — APPROVED** [review]

**Pull request closed without merging**